onnxruntime/core/providers/cpu/math/element_wise_ops.cc

```diff
@@ -273,25 +275,89 @@ Status Sqrt<T>::Compute(OpKernelContext* ctx) const {
   return Status::OK();
 }
 
-template <typename T>
-Status Pow<T>::Compute(OpKernelContext* context) const {
-  const Tensor& Y = *context->Input<Tensor>(1);
-  std::function<void(EigenVectorMap<T>, ConstEigenVectorMap<T>, T)> input1scalar =
-      [](EigenVectorMap<T> output, ConstEigenVectorMap<T> input0, T input1) { output = Eigen::pow(input0.array(), input1); };
-  if (Y.Shape().Size() == 1) {
-    T value = *Y.Data<T>();
-    if (value == 2.0) {
-      input1scalar = [](EigenVectorMap<T> output, ConstEigenVectorMap<T> input0, T) { output = Eigen::square(input0.array()); };
-    } else if (value == 3.0) {
-      input1scalar = [](EigenVectorMap<T> output, ConstEigenVectorMap<T> input0, T) { output = Eigen::cube(input0.array()); };
-    }
+namespace pow_internal {
+
+template <typename T, typename E>
+void PowImpl(OpKernelContext* context, const Tensor& X, const Tensor& Y) {
+  TBroadcaster<T, E> bc{X, Y};
+  Tensor* const output_tensor = context->Output(0, bc.GetOutputShape());
+  TBroadcastOutput<T> output{bc.GetSpanSize(), *output_tensor};
+
+  // Scalar base
+  auto input0scalar = [](gsl::span<T> output, const T& X, gsl::span<const E> Y) {
+    std::transform(Y.cbegin(), Y.cend(), output.begin(),
+                   [X](E y) {
+                     return static_cast<T>(std::pow(X, y));
+                   });
+  };
+
+  auto input1scalar = [](gsl::span<T> output, gsl::span<const T> X, const E& Y) {
+    std::transform(X.cbegin(), X.cend(), output.begin(),
+                   [Y](T x) {
+                     return static_cast<T>(std::pow(x, Y));
+                   });
+  };
+
+  auto general = [](gsl::span<T> output, gsl::span<const T> X, gsl::span<const E> Y) {
+    std::transform(
+        X.cbegin(), X.cend(), Y.cbegin(), output.begin(),
+        [](T x, E y) {
+          return static_cast<T>(std::pow(x, y));
+        });
+  };
+
+  BroadcastLoopSpan(bc, output, input0scalar, input1scalar, general);
+}
+
+template <typename B>
+void DispatchOnBase(OpKernelContext* context, const Tensor& X, const Tensor& Y) {
+  namespace on = ONNX_NAMESPACE;
+  switch (Y.GetElementType()) {
+    case on::TensorProto_DataType_INT32:
+      PowImpl<B, int32_t>(context, X, Y);
+      break;
+    case on::TensorProto_DataType_INT64:
+      PowImpl<B, int64_t>(context, X, Y);
+      break;
+    case on::TensorProto_DataType_FLOAT:
+      PowImpl<B, float>(context, X, Y);
+      break;
+    case on::TensorProto_DataType_DOUBLE:
+      PowImpl<B, double>(context, X, Y);
+      break;
+    default:
+      assert(false);
   }
```

Comment:
Can this return Status with an error message saying an unexpected type was found?